src/Fractionalizer.sol

```diff
@@ -0,0 +1,357 @@
+    mapping(address => mapping(uint256 => uint256)) claimAllowance;
+
+    address immutable tokenImplementation;
+
+    function initialize(IPNFT _ipnft, SchmackoSwap _schmackoSwap) public initializer {
+        __UUPSUpgradeable_init();
+        __Ownable_init();
+
+        ipnft = _ipnft;
+        schmackoSwap = _schmackoSwap;
+    }
+
+    modifier notClaiming(uint256 fractionId) {
+        if (address(fractionalized[fractionId].paymentToken) != address(0)) {
+            revert("already in claiming phase");
+        }
+        _;
+    }
+
+    constructor() {
+        tokenImplementation = address(new FractionalizedTokenUpgradeable());
+        _disableInitializers();
+    }
+
+    function setFeeReceiver(address _feeReceiver) public onlyOwner {
+        if (_feeReceiver == address(0)) {
+            revert ToZeroAddress();
+        }
+        feeReceiver = _feeReceiver;
+    }
+
+    function setReceiverPercentage(uint256 fractionalizationPercentage_) external onlyOwner {
+        fractionalizationPercentage = fractionalizationPercentage_;
+    }
+
+    function balanceOf(address owner, uint256 fractionId) public view returns (uint256) {
+        return fractionalized[fractionId].tokenContract.balanceOf(owner);
+    }
+
+    function totalSupply(uint256 fractionId) public view returns (uint256) {
+        return fractionalized[fractionId].tokenContract.totalSupply();
+    }
+
+    /**
+     * @notice
+     * @param ipnftId          uint256  the token id on the origin collection
+     * @param fractionsAmount  uint256  the initial amount of fractions issued
+     * @param agreementCid    bytes32  a content hash that identifies the terms underlying the issued fractions
+     */
+
+    function fractionalizeIpnft(uint256 ipnftId, uint256 fractionsAmount, string calldata agreementCid) external returns (uint256 fractionId) {
+        if (ipnft.totalSupply(ipnftId) != 1) {
+            revert("IPNFT supply must be 1");
+        }
+        if (ipnft.balanceOf(_msgSender(), ipnftId) != 1) {
+            revert("only owner can initialize fractions");
+        }
+        string memory ipnftSymbol = ipnft.symbol(ipnftId);
+        if (bytes(ipnftSymbol).length == 0) {
+            revert("ipnft needs a symbol to fractionalize");
+        }
+
+        fractionId = uint256(keccak256(abi.encodePacked(_msgSender(), ipnftId)));
+
+        // ensure we can only call this once per sales cycle
+        if (address(fractionalized[fractionId].originalOwner) != address(0)) {
+            revert("token is already fractionalized");
+        }
+
+        // https://github.com/OpenZeppelin/workshops/tree/master/02-contracts-clone
+        FractionalizedTokenUpgradeable fractionalizedToken = FractionalizedTokenUpgradeable(Clones.clone(tokenImplementation));
+        string memory name = string(abi.encodePacked("Fractions of IPNFT #", Strings.toString(ipnftId)));
+        string memory symbol = string(string(abi.encodePacked(ipnftSymbol, "-FAM")));
```

Comment:
why do we need additional string here?

Reply:
well, that's imo a requirement? The IPNFT has a symbol and its fractions have a symbol that derives from it.